### PR TITLE
feat: new experiment table filter

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[WEB-983]
- still working on it
- see the preview in `det/test`

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by alexjdetermined** [comment]
Layout
* The "Some/All of the following conditions are true" label should not be visible outside of a condition group (should only show up in a condition group and not at root level).
* Replace "trashcan" icon with "X" icon (see "Sort" PR for alignment)
* Move "show archived" toggle to below the button group left aligned (Toggle then label)

Copy changes
* Change condition group header copy from "Some | All of the following conditions are true" to "Any | All of the following are true…"
* change button copy from "clear all" to "Reset". AND....... When clicked, reset should clear all and close menu.
* Align copy for the "Add condition field" button in "+" to match main buttons "+ Add condition" (remove "field")

Bugs
* cant type to filter/select a column or value in select boxes
* tags type ahead "contains" j cause error
* tags, state, user, resource pool, searcher type should be single/multi select for "is, is not, is any of"

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
+import { Observable, observable } from 'micro-observables';
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterFormStore.ts:1]: if i use state, i need to have logic inside components. Without observable, it doesnt update state in the FilterFormStore

File: webui/react/src/pages/F_ExpList/glide-table/TableActionBar.module.scss [hunk 2]
```diff
@@ -1,6 +1,4 @@
 .base {
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/F_ExpList/glide-table/TableActionBar.module.scss:1]: accidentally removed while debugging. added back

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 3]
```diff
@@ -223,6 +229,22 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
       docTitle={project.id === 1 ? 'Uncategorized Experiments' : 'Project Details'}
       id="projectDetails">
       <>
+        <Space>
+          <TableFilter formStore={formStore} loadableColumns={projectColumns} />
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:233]: i think its double scrollbar issue. This should be fixed by Guangqing's PR

File: webui/react/src/components/FilterForm/TableFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    setOpen(newOpen);
+  };
+
+  const onHidePopOver = () => {
+    setOpen(false);
+  };
+
+  return (
+    <div>
+      <Popover
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/components/FilterForm/TableFilter.tsx:32]: According to [antd doc](https://ant.design/components/tooltip?theme=dark), 

>The tip is shown on mouse enter, and is hidden on mouse leave. The Tooltip doesn't support complex text or operations. 

so i'll leave it as it is. If this should be replaced, that would be a separate PR

**Comment by julian-determined-ai** [comment]
noting that I see a bug not respecting the project ID when the page first loads within a project I do see only the experiments within a project. However, if I add a filter with an OR statement that results to true then I see all experiments. 
It is not yet clear to me if this is a problem on the frontend or backend. 

253137F0-17AB-40CA-997C-4ABAEDEDC5AB
8232133D-9A33-4293-8EFC-7B559638D34A

**Comment by ClaireNeveu** [comment]
I'm adding one of our tech leads as a required reviewer on this since it adds 1,500+ lines of code across 12 new files.

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    const filterSet: Readonly<FilterFormSet> = this.#formset.get();
+    const filterGroup = filterSet.filterGroup;
+    const ans = this.#getFormById(filterGroup, id);
+    if (ans && ans.kind === FormKind.Group && Object.values(Conjunction).includes(conjunction)) {
+      ans.conjunction = conjunction;
+      this.#formset.set({ filterGroup, showArchived: filterSet.showArchived });
+    }
+  }
+
+  public setFieldValue(id: string, value: FormFieldValue): void {
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterFormStore.ts:124]: you should be able to merge `setFieldValue`, `setFieldColumnName` and `setFieldOperator` into a single function that takes a partial update, no?

**Review by ashtonG — COMMENTED** [review]
this is really good. I had some initial thoughts for you while i start testing.

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts (hunk 5 above)
Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterFormStore.ts:124]: it used to a single function, but i didnt look good, so i split into different functions

File: webui/react/src/components/FilterForm/components/type.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+export type FormKind = ValueOf<typeof FormKind>;
+
+export type FormFieldValue = string | number | null;
+
+export type FormField = {
+  readonly id: string;
+  readonly kind: typeof FormKind.Field;
+  columnName: V1ProjectColumn['column'];
+  location: V1LocationType;
+  type: V1ColumnType;
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/type.ts:18]: do we need `location` and `type` in the field?

Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/type.ts:18]: Filter API requires these

File: webui/react/package.json [hunk 7]
```diff
@@ -68,8 +68,8 @@
     "smoothscroll-polyfill": "^0.4.4",
     "sprintf-js": "^1.1.2",
     "throttle-debounce": "^5.0.0",
-    "undici": "^5.22.0",
```
Comment (ashtonG) [inline webui/react/package.json:71]: add this back please!

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+export class FilterFormStore {
+  #formset: WritableObservable<FilterFormSet>;
+
+  constructor(data?: Readonly<FilterFormSet>) {
+    this.init(data);
+  }
+
+  public init(data?: Readonly<FilterFormSet>): void {
+    this.#formset = observable<FilterFormSet>(structuredClone(data ? data : INIT_FORMSET));
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterFormStore.ts:49]: we'll need a shim for this for testing because this global isn't in our minimum version of node.

Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterFormStore.ts:48]: i have node update PR #6759. when its merged, i'll rebase commits

File: webui/react/src/components/FilterForm/components/FilterField.tsx [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+  const resourcePools = Loadable.getOrElse([], useObservable(clusterStore.resourcePools));
+
+  const currentColumn = columns.find((c) => c.column === field.columnName);
+  const isSpecialColumn = (SpecialColumnNames as ReadonlyArray<string>).includes(field.columnName);
+
+  const [fieldValue, setFieldValue] = useState<FormFieldValue>(field.value);
+
+  // use this function to update field value
+  const updateFieldValue = (fieldId: string, value: FormFieldValue, isDounce = false) => {
+    if (isDounce) {
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterField.tsx:81]: consider a clearer name for `isDounce`, maybe `debounceUpdate`?

**Review by ashtonG — COMMENTED** [review]
are we not storing the filter state in the url or settings?

**Comment by keita-determined** [comment]
> Looks great.
> 
> One small bug, the "icon with tooltips" icon does not have a tooltip on it (area in screenshot). image

i think this comment should be in a different PR?

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+      return key === 'id' ? undefined : value;
+    };
+    return this.#formset.select((formset) => {
+      const newFormSet: FilterFormSetWithoutId = JSON.parse(JSON.stringify(formset, replacer));
+      return newFormSet;
+    });
+  }
+
+  #getFormById(filterGroup: FormGroup, id: string): FormField | FormGroup | undefined {
+    const recur = (form: FormGroup | FormField): FormGroup | FormField | undefined => {
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterFormStore.ts:68]: changed `recur` to `traverse`

**Comment by keita-determined** [comment]
> are we not storing the filter state in the url or settings?

added

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 11]
```diff
@@ -81,6 +96,19 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [page, sortString]);
 
+  useEffect(() => {
+    let ignore = false;
```
Comment (julian-determined-ai) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:100]: **nit**: @keita-determined as per discussion could you add a comment on why you implemented the useEffect this way, perhaps document the bug/issue this solves.

**Review by julian-determined-ai — APPROVED** [review]
This was great work 🎉

**Comment by alexjdetermined** [comment]
1. User list should be alpha sorted
2. User list should allow for multi select “is any of, is none of”
5. Need a “No matching results” view instead of "empty state"
6. Show table loading state if applied filter takes too long to load (currently on Slow 3g i can apply a filter and close menu before the filter has actually filtered the table)
7. Non numeric field type operations should be: [ contains, does not contain, is, is not, is empty, is not empty ]
9. Clicking row header should set focus to the “value” input filed (third one in) for new row that was added
10. Optional fields should have [ is empty, is not empty ] operators added to the list (checkpoint size, forked from, all “metric” and “hp” columns,
11. Cant get duration field to filter. not sure if an operator issue or if the col isnt hooked up to filtering? (Consider removing and addressing "value" input later in separate ticket)
14. “Escape key” should allow me to exit the "field im focused in” if im in a field then also allow me to close the menu via escape key.
15. Name and desc “contains” and “does not contain” shouldn't be case sensitive.
16. “user” should be validated to only the options for the experiments in this project and not “every user”
17. hitting “enter” on the “value” field should add new condition row
18. Filter button should show how many conditions are applied (e.g. Filter (4), this should be for the amount of “conditions” applied in total across any condition groups as well)
21. Remove “+" icon from condition group “Add menu”
22. Remove extra padding from “FilterGroup_groupCard_bead7 undefined” and extra bottom padding from FilterGroup_base_bead7 FilterGroup_baseRoot_bead7. Explore combining the two divs in the condition group into some sort of “condition group header” and applying the padding there to address padding issues (8px left and 8px top padding)
23. Clear “incomplete” conditions and groups when exiting the menu. “on close clear the garbage”

**Comment by keita-determined** [comment]
1. User list should be alpha sorted
5. Need a “No matching results” view instead of "empty state"
6. Show table loading state if applied filter takes too long to load (currently on Slow 3g i can apply a filter and close menu before the filter has actually filtered the table)
7. Non numeric field type operations should be: [ contains, does not contain, is, is not, is empty, is not empty ]
21. Remove “+" icon from condition group “Add menu”
22. Remove extra padding from “FilterGroup_groupCard_bead7 undefined” and extra bottom padding from FilterGroup_base_bead7 FilterGroup_baseRoot_bead7. Explore combining the two divs in the condition group into some sort of “condition group header” and applying the padding there to address padding issues (8px left and 8px top padding)


I think these are what i can finish in this PR, but other things requires something else like backend change etc

**Comment by alexjdetermined** [comment]
For the "No matching results" state
1. Use the "Empty" component from the "States without data" section in the design kit
2. Replace the table with that component (ideally matching background color or "no background color" but tbd on what the component supports)
3. Action bar needs to still be visible to modify filters
4. Change copy to "No matching results"

image

**Comment by keita-determined** [comment]
>Show table loading state if applied filter takes too long to load (currently on Slow 3g i can apply a filter and close menu before the filter has actually filtered the table)

@alexjdetermined 

This get tricky due to the combination of sort and filter (`resetPagination` shouldnt be used if i understand it correctly after sort). i think we can refactor this later

**Comment by ashtonG** [comment]
@keita-determined I think we should reset pagination when the filter changes -- Pagination occurs after filtering, so the pages before filtering don't line up to the pages after filtering. Resetting the pagination would ensure that users don't get lost or inadvertently navigate beyond the result set.

File: webui/react/src/components/FilterForm/components/FilterGroup.tsx [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            // and target item's level is over 2
+            // 2 is the max depth
+            (item.form.children.filter((c) => c.kind === FormKind.Group).length === 0 ? 0 : 1) +
+              level <
+              2
+          );
+        }
+        return true;
+      }
+      return false;
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterGroup.tsx:69]: very tiny nit: maybe instead use a boolean statement to represent this so we understand why we're returning `true` or `false` near the end?
```suggestion
      return isOverCurrent && (item.form.kind !== FormKind.Group || (
            // cant self dnd
            group.id !== item.form.id &&
            // cant dnd in self childrens group
            item.form.children.filter((c) => c.id === group.id).length === 0 &&
            // cant dnd with deeper than 2 level group
            level < 2 &&
            // cant dnd if sum of source children of group type (0 if none, 1 if children exist)
            // and target item's level is over 2
            // 2 is the max depth
            (item.form.children.filter((c) => c.kind === FormKind.Group).length === 0 ? 0 : 1) +
              level <
              2
      ))
```

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 13]
```diff
@@ -81,6 +98,21 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [page, sortString]);
 
+  useEffect(() => {
+    // useSettings load the default value first, and then load the data from DB
+    // use this useEffect to re-init the correct useSettings value when settings.filterset is changed
+    let ignore = false;
+
+    if (!ignore) {
+      const formSet: FilterFormSet = JSON.parse(settings.filterset);
```
Comment (ashtonG) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:107]: Should we use an io-ts parser here to ensure the input is what we expect?

**Review by ashtonG — APPROVED** [review]
great job here!

**Comment by keita-determined** [comment]
The reason why i think these should be fixed in different PRs

> User list should allow for multi select “is any of, is none of”

Backend currently doesn't support array type. If backend support it, frontend can support array too.

> Show table loading state if applied filter takes too long to load (currently on Slow 3g i can apply a filter and close menu before the filter has actually filtered the table)

Loading state is not used just for filter (first fetch, sort using it too), and when sort is applied, basic states (pagination etc) is reset I think we need more design definition. For example, Airtable doesn't show loading state and not pagination reset afaik. this should be fixed in a different PRs since its not just for filter. 

> Clicking row header should set focus to the “value” input filed (third one in) for new row that was added

That requires `ref` to achieve it. Its technically feasible but the code would be so complicated and it might cause bugs rom engineer perspective. Airtable doesn't support this either maybe because its the same reason.

> Cant get duration field to filter. not sure if an operator issue or if the col isnt hooked up to filtering? (Consider removing and addressing "value" input later in separate ticket)

I think backend needs to support duration and searcher type.

> “Escape key” should allow me to exit the "field im focused in” if im in a field then also allow me to close the menu via escape key.

It is not just for filtering. I think we want this feature for all popoever component. This should be fixed in a different PR as UI Kit improvement

> Name and desc “contains” and “does not contain” shouldn't be case sensitive.

backend changed required

> “user” should be validated to only the options for the experiments in this project and not “every user”

Id like to do that in a different PR

> hitting “enter” on the “value” field should add new condition row

That might make code more complicated, and cause bugs 

> Filter button should show how many conditions are applied (e.g. Filter (4), this should be for the amount of “conditions” applied in total across any condition groups as well)

I dont think showing the number of condition is not desired here because the number could be at most 50 ^ 3, so i think what we can do is that filter is applied or not

23. Clear “incomplete” conditions and groups when exiting the menu. “on close clear the garbage”

I believe some users wanna keep the structure especially filter allows users to make complicated structure.

**Comment by keita-determined** [comment]
- [x] User list should be alpha sorted
- [ ] ~~User list should allow for multi select “is any of, is none of”~~
- [x] Need a “No matching results” view instead of "empty state"
- [x] Show table loading state if applied filter takes too long to load (currently on Slow 3g i can apply a filter and close menu before the filter has actually filtered the table)
- [x] Non numeric field type operations should be: [ contains, does not contain, is, is not, is empty, is not empty ]
- [x] Clicking row header should set focus to the “value” input filed (third one in) for new row that was added
- [x] Optional fields should have [ is empty, is not empty ] operators added to the list (checkpoint size, forked from, all “metric” and “hp” columns,
- [ ] ~~Cant get duration field to filter. not sure if an operator issue or if the col isnt hooked up to filtering? (Consider removing and addressing "value" input later in separate ticket)~~
- [ ] “Escape key” should allow me to exit the "field im focused in” if im in a field then also allow me to close the menu via escape key.
- [ ] ~~Name and desc “contains” and “does not contain” shouldn't be case sensitive.~~
- [ ] “user” should be validated to only the options for the experiments in this project and not “every user”
- [x] hitting “enter” on the “value” field should add new condition row
- [ ] Filter button should show how many conditions are applied (e.g. Filter (4), this should be for the amount of “conditions” applied in total across any condition groups as well)
- [x] Remove “+" icon from condition group “Add menu”
- [x] Remove extra padding from “FilterGroup_groupCard_bead7 undefined” and extra bottom padding from FilterGroup_base_bead7 FilterGroup_baseRoot_bead7. Explore combining the two divs in the condition group into some sort of “condition group header” and applying the padding there to address padding issues (8px left and 8px top padding)
- [ ] Clear “incomplete” conditions and groups when exiting the menu. “on close clear the garbage”

File: webui/react/src/components/FilterForm/components/FilterGroup.tsx (hunk 12 above)
Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterGroup.tsx:69]: it looks concise but it doesnt narrow the `form. kind` type due to `item.form.kind !== FormKind.Group`

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 13 above)
Comment (keita-determined) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:107]: does useSettings store object (i know it stores array)? IIRC, it doesnt store objects

Comment (ashtonG) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:103]: I don't think so -- you might not be able to use the settings pattern we use elsewhere, but you should be able to pass the result of the json parse into the parser?

File: webui/react/src/components/FilterForm/components/FilterGroup.tsx (hunk 12 above)
Comment (keita-determined) [reply webui/react/src/components/FilterForm/components/FilterGroup.tsx:69]: you are right. i probably replaced a wrong place

**Comment by alexjdetermined** [comment]
Out of scope for this PR
1. User multi-select
2. User value list validated to within this project
3. Row header click through to focus
4. Duration/searcher type filter support
5. Escape key/keyboard support for menus/inputs
6. Hitting enter to add another condition below last one


In scope for this PR from comments above
1. Loading ( i see it checked off so i imagine this is done?)
2. Show count of applied filter conditions 
4. Clear the trash


**Re: filter condition count** 
"I dont think showing the number of condition is not desired here because the number could be at most 50 ^ 3, so i think what we can do is that filter is applied or not" 
* not sure im following on the 50^3. The count is the number of filters conditions applied and should realistically be "1-5". Knowing if the table is being filtered by 1 condition or 10 is a meaningful heuristic of context for whats in the table without having to open the menu to see whats applied. 

**Re: clear the trash**
"I believe some users wanna keep the structure especially filter allows users to make complicated structure."
* the filter menu should reflect only things that are applied and users shouldnt have to parse to know whats active and not.

**Comment by keita-determined** [comment]
>Re: filter condition count
not sure im following on the 50^3. The count is the number of filters conditions applied and should realistically be "1-5". Knowing if the table is being filtered by 1 condition or 10 is a meaningful heuristic of context for whats in the table without having to open the menu to see whats applied.

Each group can have up to 50 conditions and filterset allows to have up to 2 nested groups, so the maximum number of condition could be 50^3 theoretically. I dont think no one will add this many conditions, but I just wanted to clarify the possibility  


> Re: clear the trash
the filter menu should reflect only things that are applied and users shouldnt have to parse to know whats active and not

ok will implement it. the trash conditions will be removed when popover is closed.

**Comment by keita-determined** [comment]
>In scope for this PR from comments above
Loading ( i see it checked off so i imagine this is done?)
Show count of applied filter conditions
Clear the trash

implemented all of above.

File: webui/react/src/components/FilterForm/components/FilterFormStore.ts [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+  public get fieldCount(): number {
+    const countFields = (form: FormGroup): number => {
+      let count = 0;
+      for (const child of form.children) {
+        count += child.kind === FormKind.Group ? countFields(child) : 1;
+      }
+      return count;
+    };
+    return countFields(this.#formset.get().filterGroup);
+  }
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterFormStore.ts:72]: This is read on every render, but should only change when the formset changes. Consider selecting this instead:
```suggestion
  public get fieldCount(): Observable<number> {
    const countFields = (form: FormGroup): number => {
      let count = 0;
      for (const child of form.children) {
        count += child.kind === FormKind.Group ? countFields(child) : 1;
      }
      return count;
    };
    return this.#formset.select(({ filterGroup }) => countFields(filterGroup));
  }
```

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 15]
```diff
@@ -63,6 +70,16 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
   const [projectColumns, setProjectColumns] = useState<Loadable<ProjectColumn[]>>(NotLoaded);
+  const [isOpenFilter, setIsOpenFilter] = useState<boolean>(false);
+  const filtersString: string = JSON.stringify(useObservable(formStore.jsonWithoutId));
+  const rootFilterChildren = useObservable(formStore.formset).filterGroup.children;
+
+  const onIsOpenFilterChange = useCallback((newOpen: boolean) => {
+    setIsOpenFilter(newOpen);
+    if (!newOpen) {
+      setTimeout(() => formStore.sweep(), 500);
+    }
```
Comment (ashtonG) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:81]: Why do we have a `setTimeout` here?

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 16]
```diff
@@ -81,6 +98,28 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
+      } else {
+        const formset = formSetValidation.right;
+        formStore.init(formset);
+      }
+    }
+
+    return () => {
+      ignore = true;
+    };
+  }, [settings.filterset]);
```
Comment (ashtonG) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:121]: `ignore` isn't necessary here. That pattern is used when we're making an asynchronous call and it's possible that by the time the asynchronous call is complete, the closure in the effect is state. Here, we immediately check `ignore` after setting it -- it will always be false.
```suggestion
  useEffect(() => {
    // useSettings load the default value first, and then load the data from DB
    // use this useEffect to re-init the correct useSettings value when settings.filterset is changed
    const formSetValidation = IOFilterFormSet.decode(JSON.parse(settings.filterset));
    if (isLeft(formSetValidation)) {
      handleError(formSetValidation.left, {
        publicSubject: 'Unable to initialize filterset from settings',
      });
    } else {
      const formset = formSetValidation.right;
      formStore.init(formset);
    }
  }, [settings.filterset]);
```

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 15 above)
Comment (keita-determined) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:81]: to avoid showing sudden sweeps 

https://github.com/determined-ai/determined/assets/109113616/3434dbca-8172-430f-b271-3c5f6b232a96

**Comment by alexjdetermined** [comment]
**General issue**
1. Clicking on "Add condition" should set input focus to the first "column field" instead of the value field
2. Condition count should only show complete and applied conditions (e.g. if i have 4 conditions in the view but only 1 is complete/applied/filtering the table then the count should be 1 and not 4)

**Table re-loading issues**
Table should only reload when the records viewable in the table have changed. Currently the table is reloading for the following events that dont change the records in the table:
1. Table shouldn't reload when closing the filter menu as nothing new has been applied or removed at that time.
3. Removing incomplete conditions causes table to reload even though the table data hasn't changed
4. Adding a new condition/group immediately causes table to reload even though a complete condition hasn't been applied

**Comment by ClaireNeveu** [comment]
https://github.com/determined-ai/determined/assets/1002726/b68775b1-cbfa-4539-bf7b-bf0443b955d4

Pagination appears to be broken with your changes.

**Comment by keita-determined** [comment]
> Screen.Recording.2023-05-18.at.11.44.55.mov 
> Pagination appears to be broken with your changes.

thanks for reporting the issue.  I confirmed this happens in different PRs too, so the issue does not come from this PR 
ex: https://deploy-preview-6892--determined-ui.netlify.app/projects/1/experiments?f_explist_v2=on
could you make a ticket? 
note: this could be netlify issue since i couldn't reproduce it on local

**Comment by keita-determined** [comment]
> **General issue**
> 
> 1. Clicking on "Add condition" should set input focus to the first "column field" instead of the value field
> 2. Condition count should only show complete and applied conditions (e.g. if i have 4 conditions in the view but only 1 is complete/applied/filtering the table then the count should be 1 and not 4)
> 
> **Table re-loading issues** Table should only reload when the records viewable in the table have changed. Currently the table is reloading for the following events that dont change the records in the table:
> 
> 1. Table shouldn't reload when closing the filter menu as nothing new has been applied or removed at that time.
> 2. Removing incomplete conditions causes table to reload even though the table data hasn't changed
> 3. Adding a new condition/group immediately causes table to reload even though a complete condition hasn't been applied

all of them should be fixed

File: webui/react/src/components/FilterForm/components/FilterGroup.tsx [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+        setTimeout(() => {
+          scrollBottomRef?.current?.scrollIntoView({ behavior: 'smooth', block: 'end' });
+        }, 100);
+      }
+    };
+
+    const items: MenuProps['items'] = [
+      { key: FormKind.Field, label: <div>Add condition</div> },
+      {
+        disabled: !(0 <= level && level <= 1),
```
Comment (ashtonG) [inline webui/react/src/components/FilterForm/components/FilterGroup.tsx:108]: Can't this just be
```suggestion
        disabled: level > 1,
```

File: webui/react/src/pages/F_ExpList/glide-table/GlideTable.tsx [hunk 18]
```diff
@@ -294,28 +306,60 @@ export const GlideTable: React.FC<GlideTableProps> = ({
...
+      };
+
       const { bounds } = args;
       const items: MenuProps['items'] = [
-        ...placeholderMenuItems,
-        { type: 'divider' },
         ...sortMenuItemsForColumn(column, sorts, onSortChange),
+        { type: 'divider' },
+        {
+          disabled: false,
```
Comment (ashtonG) [inline webui/react/src/pages/F_ExpList/glide-table/GlideTable.tsx:334]: this is unnecessary, no?
```suggestion
```

**Review by ashtonG — COMMENTED** [review]
we're really close -- just a few style issues

File: webui/react/src/pages/F_ExpList/glide-table/GlideTable.tsx (hunk 18 above)
Comment (keita-determined) [reply webui/react/src/pages/F_ExpList/glide-table/GlideTable.tsx:334]: yep, its unnecessary. will remove it

**Comment by keita-determined** [comment]
@ashtonG fixed all suggestions

**Review by ashtonG — APPROVED** [review]
